runtime/src/iree/hal/local/executable_library_util.c

```diff
@@ -139,5 +139,65 @@ iree_zone_id_t iree_hal_executable_library_call_zone_begin(
 
   return z0;
 }
+#endif  // IREE_TRACING_FEATURES & IREE_TRACING_FEATURE_INSTRUMENTATION
+
+#if IREE_TRACING_FEATURES & IREE_TRACING_FEATURE_INSTRUMENTATION
+
+iree_status_t iree_hal_executable_library_setup_tracing(
+    const iree_hal_executable_library_v0_t* library,
+    iree_allocator_t host_allocator, tracy_file_mapping** out_file_mapping) {
+  iree_status_t status = iree_ok_status();
+#if (IREE_TRACING_FEATURES)
+  // Do we have any non empty src file contents?
+  bool contains_srs = false;
+  for (uint32_t i = 0; i < library->exports.count; ++i) {
+    if (library->exports.src_files[i].file_contents_length > 0) {
+      contains_srs = true;
+      break;
+    }
+  }
+  if (!contains_srs) {
+    return iree_ok_status();
+  }
+
+  // Alloc tracy file mapping
+  if (iree_status_is_ok(status)) {
+    status = iree_allocator_malloc(host_allocator, sizeof(**out_file_mapping),
+                                   (void**)out_file_mapping);
+  }
+  tracy_file_mapping* file_mapping = *out_file_mapping;
+  if (iree_status_is_ok(status)) {
+    status = iree_allocator_malloc(
+        host_allocator,
+        library->exports.count * sizeof(*file_mapping->file_contents),
+        (void**)&file_mapping->file_contents);
+  }
+
+  // Fill tracy mapping contents
+  if (iree_status_is_ok(status)) {
+    for (uint32_t i = 0; i < library->exports.count; ++i) {
+      file_mapping->file_contents[i].file_name =
+          library->exports.src_locs[i].path;
+      file_mapping->file_contents[i].file_name_length =
+          library->exports.src_locs[i].path_length;
+      file_mapping->file_contents[i].file_contents =
+          library->exports.src_files[i].file_contents;
+      file_mapping->file_contents[i].file_contents_length =
+          library->exports.src_files[i].file_contents_length;
+    }
+    file_mapping->file_mapping_length = library->exports.count;
+  }
+
+  // Sort file mappings
+  qsort(file_mapping->file_contents, file_mapping->file_mapping_length,
+        sizeof(*file_mapping->file_contents), tracy_file_contents_sort_cmp);
+
+  // Register file mappings
+  if (iree_status_is_ok(status)) {
+    iree_tracing_register_custom_file_contents(file_mapping);
```

Comment:
note that tracy is not the only tracing implementation - any function defined here needs to be defined as part of the full tracing suite (such as `-DIREE_TRACING_PROVIDER=console`)